File: app/src/main/java/com/example/eccard/filmesfamosos/ui/MainActivity.java
package com.example.eccard.filmesfamosos.ui;

import android.content.DialogInterface;
import android.content.Intent;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.example.eccard.filmesfamosos.R;
import com.example.eccard.filmesfamosos.data.network.AppApiHelper;
import com.example.eccard.filmesfamosos.data.network.model.MovieResult;
import com.example.eccard.filmesfamosos.ui.moviedetail.MovieDetailActivity;
import com.example.eccard.filmesfamosos.utils.EndlessRecyclerViewScrollListener;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static android.content.res.Configuration.ORIENTATION_LANDSCAPE;

public class MainActivity extends AppCompatActivity implements MoviesAdapter.OnMovieClickListener, GetMoviesFragment.GetMoviesCallbacks {

    private static final int GRID_COLUMNS_PORTRAIT = 2;
    private static final int GRID_COLUMNS_LANDSCAPE = 3;

    @SuppressWarnings("WeakerAccess")
    @BindView(R.id.rv_movies)
    RecyclerView mRecycleView;

    @SuppressWarnings("WeakerAccess")
    @BindView(R.id.pb_main)
    ProgressBar mPB;

    @SuppressWarnings("WeakerAccess")
    @BindView(R.id.tv_generic_error)
    TextView mTvGenericError;


    private MoviesAdapter moviesAdapter;
    private EndlessRecyclerViewScrollListener scrollListener;
    private AppApiHelper.MovieOrderType mCurrentMovieOrderType = AppApiHelper.MovieOrderType.POPULAR;

    private GetMoviesFragment mGetMoviewFrg;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        ButterKnife.bind(this);

        setUpViews();


        FragmentManager fm = getSupportFragmentManager();
        mGetMoviewFrg = (GetMoviesFragment) fm.findFragmentByTag(GetMoviesFragment.TAG);

        if ( mGetMoviewFrg == null){
            mGetMoviewFrg = new GetMoviesFragment();
            fm.beginTransaction().add(mGetMoviewFrg, GetMoviesFragment.TAG).commit();

            getMoviePage(EndlessRecyclerViewScrollListener.STARTING_PAGE_INDEX);
        } else {
            onMoviesResult(mGetMoviewFrg.getRetainMovies());
        }
    }

    private void getMoviePage(int startingPageIndex) {
        showLoading();
        mGetMoviewFrg.getData(startingPageIndex, mCurrentMovieOrderType);
    }

    // todo check internet conection
    private void setUpViews() {

        GridLayoutManager layoutManager;

        if(getResources().getConfiguration().orientation == ORIENTATION_LANDSCAPE) {
            layoutManager = new GridLayoutManager(this, GRID_COLUMNS_LANDSCAPE);
        } else {
            layoutManager = new GridLayoutManager(this, GRID_COLUMNS_PORTRAIT);
        }

        mRecycleView.setLayoutManager(layoutManager);
        mRecycleView.setHasFixedSize(true);
        moviesAdapter = new MoviesAdapter();
        moviesAdapter.setOnMovieClickListener(this);
        mRecycleView.setAdapter(moviesAdapter);

        scrollListener = new EndlessRecyclerViewScrollListener(layoutManager) {
            @Override
            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
                getMoviePage(page);
            }
        };

        mRecycleView.addOnScrollListener(scrollListener);
    }

    private void showMovies(){
        mRecycleView.setVisibility(View.VISIBLE);
        mPB.setVisibility(View.INVISIBLE);
        mTvGenericError.setVisibility(View.INVISIBLE);
    }

    private void showLoading(){
        mRecycleView.setVisibility(View.INVISIBLE);
        mPB.setVisibility(View.VISIBLE);
        mTvGenericError.setVisibility(View.INVISIBLE);
    }

    private void showLoadingError(){
        mRecycleView.setVisibility(View.INVISIBLE);
        mPB.setVisibility(View.INVISIBLE);
        mTvGenericError.setVisibility(View.VISIBLE);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if ( item.getItemId() == R.id.action_order_by){

            AlertDialog.Builder builderSingle = new AlertDialog.Builder(MainActivity.this);
            builderSingle.setTitle(getString(R.string.select_order));


            ArrayAdapter<CharSequence> arrayAdapter = ArrayAdapter.createFromResource
                    (MainActivity.this,
                            R.array.order_array,
                            android.R.layout.simple_list_item_1);

            builderSingle.setNegativeButton(android.R.string.cancel, new DialogInterface
                    .OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.dismiss();
                }
            });

            builderSingle.setAdapter(arrayAdapter, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {

                    AppApiHelper.MovieOrderType newOrderType;
                    if ( which == 0 ){
                        newOrderType = AppApiHelper.MovieOrderType.POPULAR;
                    }else {
                        newOrderType = AppApiHelper.MovieOrderType.TOP_RATED;

                    }

                    if (mCurrentMovieOrderType == newOrderType){

                        mRecycleView.post(new Runnable() {
                            @Override
                            public void run() {
                                mRecycleView.smoothScrollToPosition(0);
                            }
                        });

                    }else {

                        mCurrentMovieOrderType = newOrderType;

                        moviesAdapter.resetMovieResults();
                        moviesAdapter.notifyDataSetChanged();
                        scrollListener.resetState();

                        getMoviePage(EndlessRecyclerViewScrollListener.STARTING_PAGE_INDEX);

                    }


                }
            });
            builderSingle.show();

            return true;
        }else {
            return super.onOptionsItemSelected(item);
        }
    }



    @Override
    public void onMovieItemClick(MovieResult movie) {
        Intent intent = MovieDetailActivity.newIntent(MainActivity.this);
        intent.putExtra(MovieResult.class.getSimpleName(),movie);

        startActivity(intent);
    }

    @Override
    public void onMovieError(Throwable throwable) {
        showLoadingError();
    }

    @Override
    public void onMoviesResult(List<MovieResult> movies) {
        showMovies();
        moviesAdapter.appendMovieResults(movies);
        moviesAdapter.notifyDataSetChanged();

    }
}

File: app/src/main/java/com/example/eccard/filmesfamosos/ui/main/MainActivity.kt
package com.example.eccard.filmesfamosos.ui.main

import android.content.Context
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.ArrayAdapter
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.eccard.filmesfamosos.BR
import com.example.eccard.filmesfamosos.R
import com.example.eccard.filmesfamosos.data.network.api.AppApiHelper
import com.example.eccard.filmesfamosos.data.network.model.MovieResult
import com.example.eccard.filmesfamosos.databinding.ActivityMainBinding
import com.example.eccard.filmesfamosos.di.ViewModelProviderFactory
import com.example.eccard.filmesfamosos.ui.moviedetail.MovieDetailActivity
import com.example.eccard.filmesfamosos.utils.EndlessRecyclerViewScrollListener
//import kotlinx.android.synthetic.main.activity_main.*
import muxi.kotlin.walletfda.ui.base.BaseActivity
import javax.inject.Inject
import kotlin.math.roundToInt

//import android.R



class MainActivity : BaseActivity<ActivityMainBinding,MainViewModel>(), LifecycleOwner,MainNavigator {

//    private var moviesAdapter: MoviesAdapter? = null
    private var scrollListener: EndlessRecyclerViewScrollListener? = null
//    private var mCurrentMovieOrderType: AppApiHelper.MovieOrderType = AppApiHelper.MovieOrderType.POPULAR

//    private var mGetMovieFrg: GetMoviesFragment? = null


    private lateinit var mActivityMainBinding: ActivityMainBinding

    @Inject
    lateinit var factory: ViewModelProviderFactory

    private lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mActivityMainBinding = getViewDataBinding()
        mainViewModel.setNavigator(this)

        setUpViews()
    }

//    private fun getMoviePage(startingPageIndex: Int) {
//        showLoading()
////        mGetMovieFrg!!.getData(startingPageIndex, mCurrentMovieOrderType)
//    }

    private fun setUpViews() {

        showLoading()

        setupRecyclerView()

//        mainViewModel.loading.set(View.VISIBLE)

//        btn_retry.setOnClickListener { getMoviePage(EndlessRecyclerViewScrollListener.STARTING_PAGE_INDEX) }

//        val layoutManager: GridLayoutManager

//        layoutManager = GridLayoutManager(this, calculateBestSpanCount())

//        getViewDataBinding().rvMovies.layoutManager = layoutManager

//        getViewDataBinding().rvMovies.setHasFixedSize(true)
//        moviesAdapter = MoviesAdapter()
//        moviesAdapter!!.setOnMovieClickListener(this)
//        getViewDataBinding().rvMovies.adapter = moviesAdapter



//        scrollListener = object : EndlessRecyclerViewScrollListener(layoutManager) {
//            public override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView) {
//                getMoviePage(page)
//            }
//        }
//        getViewDataBinding().rvMovies.addOnScrollListener(scrollListener)
//        getViewDataBinding().rvMovies.addOnScrollListener(scrollListener!!)

        mainViewModel.getApiMovies().observe(this, Observer<List<MovieResult>> { movies ->
            mainViewModel.loading.set(View.INVISIBLE)
            if (movies.isEmpty()) {
                mainViewModel.showEmpty.set(View.VISIBLE)
            } else {
                mainViewModel.showEmpty.set(View.GONE)
                mainViewModel.addMoviesFromApi(movies)
                mainViewModel.getAdapter().notifyDataSetChanged()
            }
        })

        mainViewModel.selected.observe(this, Observer {
            it.getContentIfNotHandled()?.let {
                onSelectedMovie(it)
            }
        })

        mainViewModel.getDataBaseMovies().observe(this, Observer {
            mainViewModel.moviesFromDb = it
            Log.d("MainActivity","update moview from db" + it.toString())
//            mainViewModel.getAdapter().notifyDataSetChanged()
        })

    }

    private fun setupRecyclerView(){


        val posterWidth = mActivityMainBinding.rvMovies.context.resources.getDimension(R.dimen.img_view_recycler_view_holder_width)
        val windowManager = mActivityMainBinding.rvMovies.context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = windowManager.defaultDisplay
        val outMetrics = DisplayMetrics()
        display.getMetrics(outMetrics)
        val screenWidth = outMetrics.widthPixels.toFloat()
        val bestSpanCount = (screenWidth / posterWidth).roundToInt()

        mActivityMainBinding.rvMovies.setHasFixedSize(true)
        val layoutManager = GridLayoutManager(mActivityMainBinding.rvMovies.context,bestSpanCount)
        mActivityMainBinding.rvMovies.layoutManager = layoutManager
        mActivityMainBinding.rvMovies.setHasFixedSize(true)
        mActivityMainBinding.rvMovies.adapter =  mainViewModel.getAdapter()


        scrollListener = object : EndlessRecyclerViewScrollListener(layoutManager) {
            public override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView) {
                mainViewModel.getData(page)
            }
        }
        mActivityMainBinding.rvMovies.addOnScrollListener(scrollListener!!)
    }

    private fun showMovies() {
        mainViewModel.loading.set(View.INVISIBLE)
        mainViewModel.showEmpty.set(View.INVISIBLE)
    }

    override fun showLoading() {
        mainViewModel.loading.set(View.VISIBLE)
        mainViewModel.showEmpty.set(View.INVISIBLE)
    }

    private fun showLoadingError() {
        mainViewModel.loading.set(View.INVISIBLE)
        mainViewModel.showEmpty.set(View.VISIBLE)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_order_by) {

            val builderSingle = AlertDialog.Builder(this@MainActivity)
            builderSingle.setTitle(getString(R.string.select_order))


            val arrayAdapter = ArrayAdapter.createFromResource(this@MainActivity,
                    R.array.order_array,
                    android.R.layout.simple_list_item_1)

            builderSingle.setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.dismiss() }

            builderSingle.setAdapter(arrayAdapter) { dialog, which ->
                val newOrderType: AppApiHelper.MovieOrderType
                if (which == 0) {
                    newOrderType = AppApiHelper.MovieOrderType.POPULAR
                } else if (which == 1) {
                    newOrderType = AppApiHelper.MovieOrderType.TOP_RATED

                } else {
                    newOrderType = AppApiHelper.MovieOrderType.TOP_BOOKMARK
                }

                if (mainViewModel.mCurrentMovieOrderType === newOrderType) {

                    getViewDataBinding().rvMovies.post { getViewDataBinding().rvMovies.smoothScrollToPosition(0) }

                } else {
                    scrollListener!!.resetState()
                    mainViewModel.mCurrentMovieOrderType = newOrderType
                    mainViewModel.getFirstPage()
//                    getMoviePage(EndlessRecyclerViewScrollListener.STARTING_PAGE_INDEX)

                }
            }
            builderSingle.show()

            return true
        } else {
            return super.onOptionsItemSelected(item)
        }
    }


    override fun onMovieError(throwable: Throwable) {
        showLoadingError()
    }

    override fun getLayoutId(): Int = R.layout.activity_main

    override fun getViewModel(): MainViewModel {
        mainViewModel = ViewModelProviders.of(this,factory)
                .get(MainViewModel::class.java)
        return mainViewModel
    }

    override fun getBindingVariable(): Int = BR.viewModel


    override fun onSelectedMovie(movie: MovieResult) {
        val intent = MovieDetailActivity.newIntent(this@MainActivity)
        intent.putExtra(MovieResult::class.java.simpleName, movie)

        startActivity(intent)
    }

    override fun hideLoading() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showToast(message: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
